### PR TITLE
Add webpack 5 support to dependency-extraction-webpack-plugin

### LINKED ISSUE #26134: dependency-extraction-webpack-plugin doesn't work with Webpack 5
**Describe the bug**
Using `@wordpress/dependency-extraction-webpack-plugin` with Webpack 5 throws a build error:

```
[webpack-cli] Promise rejection: TypeError: Cannot read property 'tap' of undefined
[webpack-cli] TypeError: Cannot read property 'tap' of undefined
    at ExternalModuleFactoryPlugin.apply (.../node_modules/@wordpress/dependency-extraction-webpack-plugin/node_modules/webpack/lib/ExternalModuleFactoryPlugin.js:17:37)
    at .../node_modules/@wordpress/dependency-extraction-webpack-plugin/node_modules/webpack/lib/ExternalsPlugin.js:16:63
    at Hook.eval [as call] (eval at create (.../node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:1)
    at Hook.CALL_DELEGATE [as _call] (.../node_modules/webpack/node_modules/tapable/lib/Hook.js:14:14)
    at .../node_modules/webpack/lib/Compiler.js:958:23
    at Hook.eval [as callAsync] (eval at create (.../node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (.../node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)
    at Compiler.compile (.../node_modules/webpack/lib/Compiler.js:955:28)
    at .../node_modules/webpack/lib/Compiler.js:456:12
    at Compiler.readRecords (.../node_modules/webpack/lib/Compiler.js:797:11)
```

**To reproduce**
1. Install and configure Webpack.
1. Add `@wordpress/dependency-extraction-webpack-plugin`.
2. Build.
3. See error.

**Expected behavior**
Build succeeds with no error.

**Package versions:**
- `webpack` 5.1.0
- `webpack-cli` 4.0.0
- `@wordpress/dependency-extraction-webpack-plugin` 2.7.0

**Additional context**
Webpack 5 is the current stable version of Webpack. This and other `@wordpress` packages should work with it.

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes #26134. Adds support for webpack 5 (while keeping webpack 4 compat) to `dependency-extraction-webpack-plugin`. Spinoff from #26382. Description of changes I had to make follows.

**Remove type annotations, commit a hand-crafted `types.d.ts`**
After discussion with @sirreal in https://github.com/WordPress/gutenberg/pull/26382#discussion_r523049421, I decided to stop typechecking the plugin source code with TypeScript. Webpack types are not complete enough to make it work, and making a plugin that supports both webpack 4 and 5 is an impossible task. To make types available for consumers of the plugin, I added a hand-crafted `types.d.ts` file that describes the plugin's public interface, i.e., the constructor and its options. And it works in my Atom IDE:

Screenshot 2020-12-04 at 21 53 22

**Stop using the `webpack-sources` package if possible**
Depending on and importing the `webpack-sources` package is legacy now. A plugin can access the APIs on a `webpack.sources` property of the `webpack` package export.

**Descend into concatenated modules**
The `ExternalModule` instances can now be concatenated into a common module. We need to descend into `ConcatenatedModule.modules` array to find all externals.

**Don't read compiler options before defaults were applied**
In webpack 5, the default options are not applied yet when `Plugin.apply` is called. Postpone reading the `output.filename` option to the `emit` hook where the value is used. See also: https://webpack.js.org/blog/2020-10-10-webpack-5-release/#new-plugin-order

Then there are changes that are in unit tests only:

**Change how `ExternalModule` instances are identified**
In webpack 5, `ExternalModule` instanced no longer have the `external` boolean flag. We need to identify them by the `externalType` field presence.

**Sort modules by `.identifier()` rather than implicitly by `.toString()`**
Produces more stable sorting of the modules, preventing test failures due to implementation details changes. In webpack 5, the `toString()` output includes a `debugId` value, which is an incremental counter incremented on each `Module` instantiation. Looking at `.identifier()` produces more stable sorting key.

**Remove the `with-externs` unit test that depends on webpack internals**
There was a `with-interns` unit test that checked for behavior of two `ExternalsPlugin` instances, one from config, one from this plugin, that match the same request. But this tests webpack internals rather than anything else.

In webpack 4, the module factories were processed by a `SyncWaterfallHook`, using the _last_ factory value produced by the hook for a given request. In webpack 5, the hook is an `AsyncSeriesBailHook`, and it bails on the _first_ plugin returning a valid factory.

So, even if two conflicting `ExternalsPlugin` instances were declared in the same order, the behavior is different in v4 vs v5.

**How to test:**
The full webpack 5 upgrade is stacked on top of this PR in #26382, and it builds cleanly including all unit and e2e tests.

This PR alone still uses webpack 4 with the updated `dependency-extraction-webpack-plugin`. And is also green.

### CODE REVIEW

**Comment by jsnajdr** [comment]
@gziolo @sirreal This is still a draft. I'll make the PR official after the webpack 5 upgrade PR (#26382) that's stacked on top of it one starts working correctly (there are still some unit test failures) and actually proves that the webpack 5 support works 🙂 I also plan to update the PR description so that reviewers know what I changed and why.

**Comment by jsnajdr** [comment]
Ready for review now 🎉

**Comment by gziolo** [comment]
Great work, it's very complex set of changes. It might take me some time to review but it's now on my list 👍

**Comment by jsnajdr** [comment]
> it's very complex set of changes

I did my best to split the patch into small independent commits and explain in detail what they do and why. Let me know if there's something that still looks confusing.

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 1]
```diff
@@ -3,10 +3,8 @@
  */
 const { createHash } = require( 'crypto' );
 const path = require( 'path' );
-const { ExternalsPlugin } = require( 'webpack' );
-const { RawSource } = require( 'webpack-sources' );
-// Ignore reason: json2php is untyped
-// @ts-ignore
+const webpack = require( 'webpack' );
+const { RawSource } = webpack.sources || require( 'webpack-sources' );
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:7]: How about we document it in the code:
```suggestion
// With webpack 5 it uses `sources` field but for webpack 4 it has to fallback to `webpack-sources` package. 
const { RawSource } = webpack.sources || require( 'webpack-sources' );
```

or something like that.

**Review by gziolo — APPROVED** [review]
I'm not an expert in how TypeScript integration should look like but your explanation makes a lot of sense. I would defer to @sirreal who typed this package before. Code wise, everything seems to work as necessary - all test pass. Well, at least for webpack 4 that it still uses. However, I believe that it works with webpack 5 as well since it was extracted from the PR where the upgrade is handled.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:8]: Comment added 🙂

**Pull request merged**